### DISCUSSION TITLE
Influxdb tag binding failure and Write to InfluxDB failed http error 404

### DISCUSSION THREAD

**Discussion by zekeb77** [tagyoureit/nodejs-poolController#907] [Q&A] [answered]
So, I started a few days ago setting up a pool controller.  I've gotten worked through everything(except relay card, waiting on Amazon) but I cannot get InfluxDB connected.  I have an OpenHAB installation that's already writing to InfluxDB, so I thought great I'll just point Pool Controller to the same install.  I created a new DB named pool, created a user and pw, etc.  But no matter what I do Pol Controller will not connect.  

InfluxDB is version 1.7.9
Pool Controller is current revision as cloned from GitHub.

So where do I even start to troubleshoot this?  I've googled everything I can think to google, which led me to checking the influxdb config file and verifying that http is enabled and "bound" on port 8086, which it is.  I can connect using Chronograpf from a remote PC, but not pool controller.

I think there's a LOT I'm missing, heck I can't even find a log file for InfluxDB...  :(  As far as I can tell it's configured to be logging, but a query in Chronograpf returns nothing from the logs.

**Comment by tagyoureit** [comment]
Did you see these instructions? https://github.com/tagyoureit/nodejs-poolController/wiki/Bindings-Integrations#influx

You can also enable/disable the interface in the settings menu but you can fine tune with the json.

**Comment by zekeb77** [comment]
I did see that page, but my config.json file "interfaces" section doesn't look exactly like that, which is one of many things that I wasn't sure about.  My config.json file doesn't have the "batch size" line, but everything else is the same, except I "think" there's one more line at the end of the InfluxDB section, I can check later today when I get home.  I added my user credentials to the file, and all other lines are the same as in the instructions at that link.

I've enabled/disabled InfluxDB in the settings menu quite a bit trying different things, so I've gotten really familiar with that lol.

That being said, I'm not sure what I would fine tune in the json under /web/interfaces but I've looked at it trying to see if anything jumped out at me.

I've run netstat -a on my PC running InfluxDB, and it shows listening on 0.0.0.0:8086, as well as I can connect to InfluxDB from another PC using Chronograpf and the creds I created for my pool db, so I think that's telling me I have the InfluxDB side set up correctly?




________________________________

**Comment by tagyoureit** [comment]
What errors are you seeing in the logs?  (Look in the ./logs directory).

**Comment by zekeb77** [comment]
[28/02/2024, 21:43:45] error: InfluxDB batch write failed writing 21 lines with 1 failed attempts.  404 Not Found : 404 page not found

[28/02/2024, 21:42:47] error: InfluxDB tag binding failure on chlorinator:status/@bind=data.status.desc; --> @bind=data.status.desc;  {"id":50,"disabled":false,"poolSetpoint":65,"spaSetpoint":10,"superChlorHours":8,"type":{"val":0,"name":"pentair","desc":"Pentair"},"model":{"val":9,"name":"aquarite-925","desc":"AquaRite 925","capacity":25000,"chlorinePerDay":0.98,"chlorinePerSec":0.000011342592592592592},"body":{"val":0,"name":"pool","desc":"Pool"},"saltRequired":0,"name":"IntelliChlor1","isActive":true,"superChlor":false,"targetOutput":0,"currentOutput":0,"equipmentType":"chlorinator","saltTarget":3200}

[28/02/2024, 21:42:47] error: InfluxDB value not bound

[28/02/2024, 21:44:40] silly: Batching influx circuit inverse data point circuits,id=6,name=Pool,type=Pool isOn=F 1709178279602000000)
[28/02/2024, 21:44:40] silly: Batching influx circuit data point circuits,id=6,name=Pool,type=Pool isOn=T,isOnVal=1i 1709178280602000000
[28/02/2024, 21:44:40] silly: Batching influx body data point bodyTemps,body=Pool,heatMode=Off,heatStatus=Off Pool.setPoint=0i 1709178280605000000
[28/02/2024, 21:44:41] silly: Batching influx virtualCircuit inverse data point circuits,id=247,name=Pool/Spa,type=Pool/Spa isOn=F 1709178280437000000)
[28/02/2024, 21:44:41] silly: Batching influx virtualCircuit data point circuits,id=247,name=Pool/Spa,type=Pool/Spa isOn=T,isOnVal=1i 1709178281437000000
[28/02/2024, 21:44:41] silly: Batching influx filter data point filter,body=Pool,filterType=Sand,id=1,name=Filter\ 1,units=Pounds\ per\ Sqare\ Inch isOn=T,isOnVal=1i 1709178281441000000
[28/02/2024, 21:44:42] error: InfluxDB tag binding failure on chlorinator:status/@bind=data.status.desc; --> @bind=data.status.desc;  {"id":50,"disabled":false,"poolSetpoint":65,"spaSetpoint":10,"superChlorHours":8,"type":{"val":0,"name":"pentair","desc":"Pentair"},"model":{"val":9,"name":"aquarite-925","desc":"AquaRite 925","capacity":25000,"chlorinePerDay":0.98,"chlorinePerSec":0.000011342592592592592},"body":{"val":0,"name":"pool","desc":"Pool"},"saltRequired":0,"name":"IntelliChlor1","isActive":true,"superChlor":false,"targetOutput":65,"currentOutput":65,"equipmentType":"chlorinator","saltTarget":3200}
[28/02/2024, 21:44:42] error: InfluxDB value not bound

**Reply by zekeb77** [reply]
That's a few samples of what I'm seeing.  They repeat every so often.

**Reply by zekeb77** [reply]
And this is my InfluxDB section of my config.json file:

"influxDB": {
        "name": "InfluxDB",
        "type": "influx",
        "enabled": true,
        "fileName": "influxDB.json",
        "options": {
          "version": 1,
          "protocol": "http://",
          "host": "192.168.1.4",
          "port": 8086,
          "username": "pool",
          "password": "pool",
          "database": "pool",
          "retentionPolicy": "autogen",
          "tags": [
            {
              "name": "sourceIP",
              "value": "@bind=webApp.ip();"
            },
            {
              "name": "sourceApp",
              "value": "njspc"
            }
          ]
        },

**Reply by zekeb77** [reply]
And for another test I found this command and tried it after a little more Google-Fu.

curl -XPOST 192.168.1.4:8086/api/v2/query -sS


![image](https://github.com/tagyoureit/nodejs-poolController/assets/161677507/ee44c847-c659-4ecc-a086-3bf7e30abb60)


Then after a touch more reading, I found/tried this command with the respective results:  (I changed the IP, and db to pool, and nothing else)

curl -G 'http://192.168.1.4:8086/query?pretty=true' --data-urlencode "db=pool" --data-urlencode "q=SELECT \"value\" FROM \"cpu_load_short\" WHERE \"region\"='us-west'"

{
    "results": [
        {
            "statement_id": 0
        }
    ]
}

curl -G 'http://192.168.1.4:8086/query?pretty=true' --data-urlencode "db=_internalmonitor" --data-urlencode "q=SELECT \"value\" FROM \"cpu_load_short\" WHERE \"region\"='us-west'"
{
    "results": [
        {
            "statement_id": 0,
            "error": "database not found: _internalmonitor"
        }
    ]
}

(On this one I was guessing at the internal db name, too lazy to go look, so it failed.  But the point it, it hit InfluxDB as shown below, so I believe that's telling me InfluxDB is at least open on port 8086 and accepting queries)

And that caused these hits to my InfluxDB:

![image](https://github.com/tagyoureit/nodejs-poolController/assets/161677507/74483d1a-7d33-4a35-8535-4a2a7a77e831)

**Comment by tagyoureit** [comment]
Your chlorinator data isn't including a status.  The code is looking for a value in `data.status.desc` but the JSON doesn't include it.  Is your chlorinator talking properly to njsPC?

First thing you can try is commenting out the line that will send this status.  In `./web/bindings/influxDB.json` change lines 935/936 to be the following:

```json
            {
              "_name": "status",
              "_value": "@bind=data.status.desc;"
            },
```
This will effectively ignore them for now.

See if any errors come up.   This may be causing the values to not be set.  The other thing to note is the default batch size is 100 values so nothing will be sent to influx until 100 values have queued up.  

You can add the batch size in the options section as shown below:
```json
"influxDB": {
  "name": "InfluxDB",
  "type": "influx",
  "enabled": true,
  "fileName": "influxDB.json",
  "options": {
    "version": 1,
    "protocol": "http://",
    "host": "192.168.1.4",
    "port": 8086,
    "username": "pool",
    "password": "pool",
    "database": "pool",
    "retentionPolicy": "autogen",
    "tags": [
      {
        "name": "sourceIP",
        "value": "@Bind=webApp.ip();"
      },
      {
        "name": "sourceApp",
        "value": "njspc"
      }
    ],
    "batchSize": 100
  },
```

**Reply by zekeb77** [reply]
Well it wasn't, but as of yesterday it is(chlorinator).  I got it connected yesterday, so I can check tonight and see if that changed anything.

I didn't understand that about the batch size, good to know.  I would assume I could lower that number for troubleshooting purposes so it tries to write more often?

**Reply by tagyoureit** [reply]
> I didn't understand that about the batch size, good to know. I would assume I could lower that number for troubleshooting purposes so it tries to write more often?

Yup!  Batching is very effective.  Frequent writes slow down both systems but by all means you can adjust it.

**Reply by zekeb77** [reply]
Ok so I added the batch size line and set it to 20 for fun.  

When I change settings, etc, I see log entries saying it batched data to InfluxDB, but nothing shows up in InfluxDB....  ???

[04/03/2024, 17:44:14] silly: Batching influx circuit inverse data point circuits,id=2,name=Pool\ Light,type=ColorLogic isOn=T 1709595853254000000)
[04/03/2024, 17:44:14] silly: Batching influx circuit data point circuits,id=2,name=Pool\ Light,type=ColorLogic isOn=F,isOnVal=0i 1709595854254000000

I'm gonna reboot my Pi and check logs shortly after start up of the software.  I'll post those results in few mins.

**Reply by zekeb77** [reply]
So after reboot, I got these...

[04/03/2024, 17:55:54] info: Init influx interface: InfluxDB

[04/03/2024, 17:55:54] silly: Skipping INFLUX write because some data is missing with pump event on measurement pumps.

[04/03/2024, 17:55:59] silly: Batching influx temps data point ambientTemps,units=Fahrenheit airTemp=52i,airTempF=52.69,waterSensor1=69.76,waterSensor2=69.67 1709596559484000000

[04/03/2024, 17:55:59] silly: Batching influx circuit inverse data point circuits,id=6,name=Pool,type=Pool isOn=T 1709596558492000000)
[04/03/2024, 17:55:59] silly: Batching influx circuit data point circuits,id=6,name=Pool,type=Pool isOn=F,isOnVal=0i 1709596559492000000
[04/03/2024, 17:55:59] silly: Batching influx circuit inverse data point circuits,id=2,name=Pool\ Light,type=ColorLogic isOn=T 1709596558497000000)
[04/03/2024, 17:55:59] silly: Batching influx circuit data point circuits,id=2,name=Pool\ Light,type=ColorLogic isOn=F,isOnVal=0i 1709596559497000000
[04/03/2024, 17:55:59] silly: Batching influx circuit inverse data point circuits,id=3,name=Pool\ -\ Sanitize,type=Master\ Cleaner isOn=T 1709596558502000000)
[04/03/2024, 17:55:59] silly: Batching influx circuit data point circuits,id=3,name=Pool\ -\ Sanitize,type=Master\ Cleaner isOn=F,isOnVal=0i 1709596559502000000
[04/03/2024, 17:55:59] silly: Batching influx body data point bodyTemps,body=Pool,heatMode=Off,heatStatus=Off Pool=66i,Pool.setPoint=0i,PoolF=66.49 1709596559508000000

[04/03/2024, 17:55:59] silly: Skipping INFLUX write because some data is missing with pump event on measurement pumps.
[04/03/2024, 17:55:59] silly: Batching influx chlorinator data point chlorinators,id=50,name=Aquarite\ SWG,status=Very\ Low\ Salt,superChlor=false,superChlorHours=8 currentOutput=0i,poolSetpoint=65i,saltLevel=2300i,spaSetpoint=10i,target\ output=0i 1709596559519000000

[04/03/2024, 17:56:59] error: InfluxDB batch write failed writing 14 lines with 1 failed attempts.  404 Not Found : 404 page not found

04/03/2024, 18:47:45] silly: Batching influx temps data point ambientTemps,units=Fahrenheit airTemp=52i,airTempF=52.69,waterSensor1=69.6,waterSensor2=69.52 1709599665498000000

[04/03/2024, 18:48:45] error: InfluxDB batch write failed writing 1 lines with 1 failed attempts.  404 Not Found : 404 page not found

**Reply by zekeb77** [reply]
That was all occurrences of "influx" btw.

**Reply by zekeb77** [reply]
And nothing shows up in the log on InfluxDB PC.

**Reply by zekeb77** [reply]
![image](https://github.com/tagyoureit/nodejs-poolController/assets/161677507/3773a410-14c0-4342-b54d-228d47df0b56)

**Reply by rstrouse** [reply]
An http 404 means that the influx db endpoint does not exist at the url.  Post your settings because there are differences between v1 and v2 influx.

**Reply by zekeb77** [reply]
Config file in PC - InfluxdB v1

"influxDB": {
        "name": "InfluxDB",
        "type": "influx",
        "enabled": true,
        "fileName": "influxDB.json",
        "options": {
          "version": 1,
          "protocol": "http://",
          "host": "192.168.1.4",
          "port": 8086,
          "username": "pool",
          "password": "pool",
          "database": "pool",
          "retentionPolicy": "autogen",
          "tags": [
            {
              "name": "sourceIP",
              "value": "@bind=webApp.ip();"
            },
            {
              "name": "sourceApp",
              "value": "njspc"
            }
          ],
          "batchSize": 20
        },

**Reply by zekeb77** [reply]
InfluxDB config file

### Welcome to the InfluxDB configuration file.

# The values in this file override the default values used by the system if
# a config option is not specified. The commented out lines are the configuration
# field and the default value used. Uncommenting a line and changing the value
# will change the value used at runtime when the process is restarted.

# Once every 24 hours InfluxDB will report usage data to usage.influxdata.com
# The data includes a random ID, os, arch, version, the number of series and other
# usage data. No data from user databases is ever transmitted.
# Change this option to true to disable reporting.
# reporting-disabled = false

# Bind address to use for the RPC service for backup and restore.
# bind-address = "0.0.0.0:8088"

###
### [meta]
###
### Controls the parameters for the Raft consensus group that stores metadata
### about the InfluxDB cluster.
###

[meta]
  # Where the metadata/raft database is stored
  dir = "/var/lib/influxdb/meta"

  # Automatically create a default retention policy when creating a database.
  # retention-autocreate = true

  # If log messages are printed for the meta service
  # logging-enabled = true

###
### [data]
###
### Controls where the actual shard data for InfluxDB lives and how it is
### flushed from the WAL. "dir" may need to be changed to a suitable place
### for your system, but the WAL settings are an advanced configuration. The
### defaults should work for most systems.
###

[data]
  # The directory where the TSM storage engine stores TSM files.
  dir = "/var/lib/influxdb/data"

  # The directory where the TSM storage engine stores WAL files.
  wal-dir = "/var/lib/influxdb/wal"

  # The amount of time that a write will wait before fsyncing.  A duration
  # greater than 0 can be used to batch up multiple fsync calls.  This is useful for slower
  # disks or when WAL write contention is seen.  A value of 0s fsyncs every write to the WAL.
  # Values in the range of 0-100ms are recommended for non-SSD disks.
  # wal-fsync-delay = "0s"


  # The type of shard index to use for new shards.  The default is an in-memory index that is
  # recreated at startup.  A value of "tsi1" will use a disk based index that supports higher
  # cardinality datasets.
  # index-version = "inmem"

  # Trace logging provides more verbose output around the tsm engine. Turning
  # this on can provide more useful output for debugging tsm engine issues.
   trace-logging-enabled = true

  # Whether queries should be logged before execution. Very useful for troubleshooting, but will
  # log any sensitive data contained within a query.
   query-log-enabled = true

  # Validates incoming writes to ensure keys only have valid unicode characters.
  # This setting will incur a small overhead because every key must be checked.
  # validate-keys = false

  # Settings for the TSM engine

  # CacheMaxMemorySize is the maximum size a shard's cache can
  # reach before it starts rejecting writes.
  # Valid size suffixes are k, m, or g (case insensitive, 1024 = 1k).
  # Values without a size suffix are in bytes.
  # cache-max-memory-size = "1g"

  # CacheSnapshotMemorySize is the size at which the engine will
  # snapshot the cache and write it to a TSM file, freeing up memory
  # Valid size suffixes are k, m, or g (case insensitive, 1024 = 1k).
  # Values without a size suffix are in bytes.
  # cache-snapshot-memory-size = "25m"

  # CacheSnapshotWriteColdDuration is the length of time at
  # which the engine will snapshot the cache and write it to
  # a new TSM file if the shard hasn't received writes or deletes
  # cache-snapshot-write-cold-duration = "10m"

  # CompactFullWriteColdDuration is the duration at which the engine
  # will compact all TSM files in a shard if it hasn't received a
  # write or delete
  # compact-full-write-cold-duration = "4h"

  # The maximum number of concurrent full and level compactions that can run at one time.  A
  # value of 0 results in 50% of runtime.GOMAXPROCS(0) used at runtime.  Any number greater
  # than 0 limits compactions to that value.  This setting does not apply
  # to cache snapshotting.
  # max-concurrent-compactions = 0

  # CompactThroughput is the rate limit in bytes per second that we
  # will allow TSM compactions to write to disk. Note that short bursts are allowed
  # to happen at a possibly larger value, set by CompactThroughputBurst
  # compact-throughput = "48m"

  # CompactThroughputBurst is the rate limit in bytes per second that we
  # will allow TSM compactions to write to disk.
  # compact-throughput-burst = "48m"

  # If true, then the mmap advise value MADV_WILLNEED will be provided to the kernel with respect to
  # TSM files. This setting has been found to be problematic on some kernels, and defaults to off.
  # It might help users who have slow disks in some cases.
  # tsm-use-madv-willneed = false

  # Settings for the inmem index

  # The maximum series allowed per database before writes are dropped.  This limit can prevent
  # high cardinality issues at the database level.  This limit can be disabled by setting it to
  # 0.
  # max-series-per-database = 1000000

  # The maximum number of tag values per tag that are allowed before writes are dropped.  This limit
  # can prevent high cardinality tag values from being written to a measurement.  This limit can be
  # disabled by setting it to 0.
  # max-values-per-tag = 100000

  # Settings for the tsi1 index

  # The threshold, in bytes, when an index write-ahead log file will compact
  # into an index file. Lower sizes will cause log files to be compacted more
  # quickly and result in lower heap usage at the expense of write throughput.
  # Higher sizes will be compacted less frequently, store more series in-memory,
  # and provide higher write throughput.
  # Valid size suffixes are k, m, or g (case insensitive, 1024 = 1k).
  # Values without a size suffix are in bytes.
  # max-index-log-file-size = "1m"

  # The size of the internal cache used in the TSI index to store previously 
  # calculated series results. Cached results will be returned quickly from the cache rather
  # than needing to be recalculated when a subsequent query with a matching tag key/value 
  # predicate is executed. Setting this value to 0 will disable the cache, which may
  # lead to query performance issues.
  # This value should only be increased if it is known that the set of regularly used 
  # tag key/value predicates across all measurements for a database is larger than 100. An
  # increase in cache size may lead to an increase in heap usage.
  series-id-set-cache-size = 100

###
### [coordinator]
###
### Controls the clustering service configuration.
###

[coordinator]
  # The default time a write request will wait until a "timeout" error is returned to the caller.
  # write-timeout = "10s"

  # The maximum number of concurrent queries allowed to be executing at one time.  If a query is
  # executed and exceeds this limit, an error is returned to the caller.  This limit can be disabled
  # by setting it to 0.
  # max-concurrent-queries = 0

  # The maximum time a query will is allowed to execute before being killed by the system.  This limit
  # can help prevent run away queries.  Setting the value to 0 disables the limit.
  # query-timeout = "0s"

  # The time threshold when a query will be logged as a slow query.  This limit can be set to help
  # discover slow or resource intensive queries.  Setting the value to 0 disables the slow query logging.
  # log-queries-after = "0s"

  # The maximum number of points a SELECT can process.  A value of 0 will make
  # the maximum point count unlimited.  This will only be checked every second so queries will not
  # be aborted immediately when hitting the limit.
  # max-select-point = 0

  # The maximum number of series a SELECT can run.  A value of 0 will make the maximum series
  # count unlimited.
  # max-select-series = 0

  # The maximum number of group by time bucket a SELECT can create.  A value of zero will max the maximum
  # number of buckets unlimited.
  # max-select-buckets = 0

###
### [retention]
###
### Controls the enforcement of retention policies for evicting old data.
###

[retention]
  # Determines whether retention policy enforcement enabled.
  # enabled = true

  # The interval of time when retention policy enforcement checks run.
  # check-interval = "30m"

###
### [shard-precreation]
###
### Controls the precreation of shards, so they are available before data arrives.
### Only shards that, after creation, will have both a start- and end-time in the
### future, will ever be created. Shards are never precreated that would be wholly
### or partially in the past.

**Reply by zekeb77** [reply]
First few lines of InlfuxDB command line int when started.  Shows version as 1.7.9

![image](https://github.com/tagyoureit/nodejs-poolController/assets/161677507/6688eb20-fc91-4c35-bf88-d6c71addb3fd)

**Reply by zekeb77** [reply]
Are there any other pertinent settings, logs, etc?

**Reply by zekeb77** [reply]
I'm pretty sure the logs tell everything you can see here but for good measure I'm including a screenshot of the Pool Controller dashboard showing the InfluxDB interfaces config.

![image](https://github.com/tagyoureit/nodejs-poolController/assets/161677507/dcc2d747-07ab-46f4-bdca-e4812d59d665)

**Comment by tagyoureit** [comment, marked as answer]
From the client library...
> Note: Use this client library with InfluxDB 2.x. For connecting to InfluxDB 1.x see the [node-influx](https://github.com/node-influx/node-influx) client library. This library can also write/query [flux-enabled](https://docs.influxdata.com/influxdb/v1.8/administration/config/#flux-enabled--false) InfluxDB 1.8+.

Can you upgrade to 1.8 from 1.7.9?  I think that's your issue.

**Reply by zekeb77** [reply]
I don't know of any reason why I can't.  However, if I'm going to upgrade is there any good reason not to go ahead and upgrade to version 2?

**Reply by tagyoureit** [reply]
Version 2.0 uses the Flux language exclusively.  Last I checked (which was a while ago), Grafana was more limited in its ability to query and interact v2.0 compared to v1.8.  We have [sample dashboards for v1.8](https://github.com/tagyoureit/nodejs-poolController/wiki/Bindings-Integrations#sample-grafana-dashboard-examples).

**Reply by zekeb77** [reply]
Understood, I'll stick with 1.8.

**Reply by zekeb77** [reply]
Well, that was the ticket it seems.  Upgraded to v1.8.1 and now have data being written to Influxdb... Thank you for your help, and your PATIENCE!
